src/persist-client/src/impl/maintenance.rs

```diff
@@ -0,0 +1,194 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! All machines need maintenance
+//!
+//! Maintenance operations for persist, shared among active handles
+
+use crate::r#impl::compact::CompactReq;
+use crate::r#impl::gc::GcReq;
+use crate::{Compactor, GarbageCollector, Machine, ReaderId, WriterId};
+use differential_dataflow::difference::Semigroup;
+use differential_dataflow::lattice::Lattice;
+use mz_persist_types::{Codec, Codec64};
+use std::fmt::Debug;
+use timely::progress::Timestamp;
+use tokio::task::JoinHandle;
+
+#[derive(Debug, Default)]
+pub struct LeaseExpiration {
+    pub(crate) readers: Vec<ReaderId>,
+    pub(crate) writers: Vec<WriterId>,
+}
+
+/// Every handle to this shard may be occasionally asked to perform
+/// routine maintenance after a successful compare_and_set operation.
+///
+/// For one-shot operations (like registering a reader) handles are
+/// allowed to skip routine maintenance if necessary, as the same
+/// maintenance operations will be recomputed by the next successful
+/// compare_and_set of any handle.
+///
+/// Operations that run regularly once a handle is registered, such
+/// as heartbeats, are expected to always perform maintenance.
+#[must_use]
+#[derive(Debug, Default)]
+pub struct RoutineMaintenance {
+    pub(crate) garbage_collection: Option<GcReq>,
+    pub(crate) lease_expiration: Option<LeaseExpiration>,
+}
+
+impl RoutineMaintenance {
+    /// Initiates any routine maintenance necessary in background tasks
+    pub(crate) fn perform<K, V, T, D>(self, machine: &Machine<K, V, T, D>, gc: &GarbageCollector)
+    where
+        K: Debug + Codec,
+        V: Debug + Codec,
+        T: Timestamp + Lattice + Codec64,
+        D: Semigroup + Codec64,
+    {
+        let _ = self.perform_in_background(machine, gc);
+    }
+
+    /// Performs any routine maintenance necessary in background tasks. Can be awaited
+    /// to ensure all background work has completed.
+    ///
+    /// Used for testing maintenance-related state transitions deterministically
+    #[cfg(test)]
+    #[allow(dead_code)]
```

Comment:
is this still dead_code? I thought I saw a test usage of it

Reply:
only used in test for WriterMaintenance right now, but I wanted both variants to have equivalent implementations (/ foresee many future tests involving this method on both)

Reply:
pretty sure you don't need the dead_code if it's both `cfg(test)` and used in a test

Reply:
ah sorry -- WriterMaintenance::perform_awaitable is used in a test. RoutineMaintenance::perform_awaitable is not yet